### PR TITLE
87 initialize new request with dynamic forms

### PR DESCRIPTION
**Pull request by alishaevn**
related: https://github.com/scientist-softserv/webstore-component-library/pull/109

# story
- this pr allows us to press the "request this item" button on the `/browse` page and be directed to a new request
- the request page is created partially using [react json schema form](https://react-jsonschema-form.readthedocs.io/en/latest/), and partially with components from the library
- we are accounting for when a ware doesn't have a dynamic form in the `useInitializeRequest` function, but not in the ui
- pressing "initiate request" only console logs the data for now. another pr will address submitting the info to the api

# demo

https://user-images.githubusercontent.com/29032869/212176130-d0792d68-ca49-4704-9a68-0acd17501d5a.mp4

### CODE REVIEW

File: pages/browse/index.js [hunk 1]
```diff
@@ -26,6 +26,7 @@ const Browse = () => {
       <SearchBar onSubmit={handleOnSubmit} initialValue={existingQuery} />
       {isLoading
         ? (
+          // TODO(alishaevn): refactor for prop error: missing "item.id"
```
Comment (alishaevn) [inline pages/browse/index.js:29]: lint fix

File: pages/index.js [hunk 2]
```diff
@@ -17,7 +17,10 @@ const Home = () => {
   const router = useRouter()
   const { wares, isLoading, isError } = useAllWares()
   const featuredServices = configureServices({ data: wares, path: '/services' })?.slice(0, 3)
-  const handleOnSubmit = ({ value }) => router.push({ pathname: '/browse', query: { q: value } }, (value.length > 0 ? `/browse?q=${value}` : '/browse'))
+  const handleOnSubmit = ({ value }) => router.push(
+    { pathname: '/browse', query: { q: value } },
+    (value.length > 0 ? `/browse?q=${value}` : '/browse')
+  )
```
Comment (alishaevn) [inline pages/index.js:23]: lint fix

File: pages/requests/new/[ware].js [hunk 3]
```diff
@@ -38,16 +39,12 @@ const NewServiceRequest = () => {
       country: '',
       text: '',
     },
-    data: {
-      timeline: '',
-      description: '',
-      // TODO(alishaevn): does the api post function account for the supplier or does that need to be part of state?
-      suppliersIdentified: 'Yes',
-    },
```
Comment (alishaevn) [inline pages/requests/new/[ware].js:46]: this info is gathered from the react json schema form now

File: pages/requests/new/[ware].js [hunk 4]
```diff
@@ -70,52 +67,56 @@ const NewServiceRequest = () => {
...
+      <Title title={dynamicForm.name || ''} addClass='my-4' />
+      {dynamicForm.schema &&
+        <Form
+          formData={formData}
+          onChange={e => setFormData(e.formData)}
+          onSubmit={handleSubmit}
+          schema={dynamicForm.schema}
+          uiSchema={dynamicForm.uiSchema}
+          validator={validator}
+        >
```
Comment (alishaevn) [inline pages/requests/new/[ware].js:97]: this is for the react schema form

File: utils/api/requests.js [hunk 5]
```diff
@@ -72,3 +72,100 @@ export const sendMessage = ({ id, message, files }) => {
 
   posting(`/quote_groups/${id}/notes.json`, note)
 }
+
+export const useInitializeRequest = (id) => {
+  const { data, error } = useSWR(`/wares/${id}/quote_groups.json`, fetcher)
+  let dynamicForm = { name: data?.name }
```
Comment (alishaevn) [inline utils/api/requests.js:78]: some wares may not have a dynamic form. we at minimum need the name of the ware though.

File: utils/api/requests.js [hunk 6]
```diff
@@ -72,3 +72,100 @@ export const sendMessage = ({ id, message, files }) => {
...
+    dynamicForm,
+    isLoadingInitialRequest: !error && !dynamicForm,
+    isInitialRequestError: error,
+  }
+}
+
+// TODO(alishaevn): https://github.com/assaydepot...scientist_api_v2/app/serializers/scientist_api_v2/dynamic_form_serializer.rb#L39
+// update the method at the code above to return the configured schema below
+export const dynamicFormSchema = (defaultSchema) => {
+  // TODO(alishaevn): may need to account for multiple forms
```
Comment (alishaevn) [inline utils/api/requests.js:102]: discussion about this is being held [here](https://assaydepot.slack.com/archives/C03FZDALABG/p1673465184439309?thread_ts=1673045342.151399&cid=C03FZDALABG).

File: utils/api/requests.js [hunk 7]
```diff
@@ -72,3 +72,100 @@ export const sendMessage = ({ id, message, files }) => {
...
+
+  const removedProperties = [
+    'concierge_support', 'suppliers_identified', 'price_comparison', 'number_suppliers',
+    'supplier_criteria', 'supplier_confirmation'
+  ]
```
Comment (alishaevn) [inline utils/api/requests.js:107]: in the webstore we are focused on a single supplier and the mvp doesn't have concierge features, so we don't need these.

File: utils/api/requests.js [hunk 8]
```diff
@@ -72,3 +72,100 @@ export const sendMessage = ({ id, message, files }) => {
...
+
+  Object.entries(defaultSchema.properties).forEach(prop => {
+    const [key, value] = prop
+    let adjustedProperty
+
+    if (!removedProperties.includes(key)) {
+      if (value.required) {
+        requiredFields.push(key)
+        let { required, ...remainingProperties } = value
+        adjustedProperty = { ...remainingProperties }
```
Comment (alishaevn) [inline utils/api/requests.js:120]: I started off using `delete value[key]`. but for some reason that kept resulting in `requiredFields = []` and `dependencyFields = {}` on the form. (see second row in image below)
Screen Shot 2023-01-12 at 12 32 07 PM

File: utils/helpers.js [hunk 9]
```diff
@@ -0,0 +1,4 @@
+export const addDays = (date, days) => {
```
Comment (alishaevn) [inline utils/helpers.js:1]: moved this over from the component library because the default value wasn't getting set in the webstore local state.

**Review by K8Sewell — APPROVED** [review]

**Comment by alishaevn** [comment]
kait isn't a part of the softserv github org right now so it's not counting her approval as one that meets the branch protections. 
this is being corrected, force merging this pr in the mean time.

**Pull request merged**